android/app/src/main/kotlin/com/yubico/authenticator/fido/data/Session.kt

```diff
@@ -0,0 +1,116 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.yubico.authenticator.fido.data
+
+import com.yubico.yubikit.fido.ctap.Ctap2Session.InfoData
+import kotlinx.serialization.*
+
+typealias YubiKitFidoSession = com.yubico.yubikit.fido.ctap.Ctap2Session
+
+@Serializable
+data class Options(
+    val clientPin: Boolean,
+    val credMgmt: Boolean,
+    val credentialMgmtPreview: Boolean,
+    val bioEnroll: Boolean?,
+    val alwaysUv: Boolean
+)
+
+fun Map<String, Any?>.getBoolean(
+    key: String,
+    default: Boolean = false
+): Boolean = get(key) as? Boolean ?: default
+
+fun Map<String, Any?>.getOptionalBoolean(
+    key: String
+): Boolean? = get(key) as? Boolean
+
+@Serializable
+data class SessionInfo(
+    val options: Options,
+    val aaguid: ByteArray,
+    @SerialName("min_pin_length")
+    val minPinLength: Int,
+    @SerialName("force_pin_change")
+    val forcePinChange: Boolean
+) {
+    constructor(infoData: InfoData) : this(
+        Options(
+            infoData.options.getBoolean("clientPin"),
+            infoData.options.getBoolean("credMgmt"),
+            infoData.options.getBoolean("credentialMgmtPreview"),
+            infoData.options.getOptionalBoolean("bioEnroll"),
+            infoData.options.getBoolean("alwaysUv")
+        ),
+        infoData.aaguid,
+        infoData.minPinLength,
+        infoData.forcePinChange
+    )
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (javaClass != other?.javaClass) return false
+
+        other as SessionInfo
+
+        if (options != other.options) return false
+        if (!aaguid.contentEquals(other.aaguid)) return false
+        if (minPinLength != other.minPinLength) return false
+        return forcePinChange == other.forcePinChange
+    }
+
+    override fun hashCode(): Int {
+        var result = options.hashCode()
+        result = 31 * result + aaguid.contentHashCode()
+        result = 31 * result + minPinLength
+        result = 31 * result + forcePinChange.hashCode()
+        return result
+    }
+
+}
+
+@Serializable
+data class Session(
+    @SerialName("info")
+    val info: SessionInfo,
+    val unlocked: Boolean,
+    val initialized: Boolean
+) {
+    constructor(infoData: InfoData, unlocked: Boolean) : this(
+        SessionInfo(infoData), unlocked, true
+    )
+
+    companion object {
+        val uninitialized = Session(
+            SessionInfo(
+                Options(
+                    clientPin = false,
+                    credMgmt = false,
+                    credentialMgmtPreview = false,
+                    bioEnroll = null,
+                    alwaysUv = false
+                ),
+                aaguid = ByteArray(0),
+                minPinLength = 0,
+                forcePinChange = false
+            ),
+            unlocked = false,
+            initialized = false
+        )
```

Comment:
I don't like the "initialized" variable on the Session objects, that's what `null` should be for. I understand that the problem with this may be that we need to differentiate on the Flutter side between "loading" and "waiting for tap", but I think we can solve that in a better way.

We use AsyncData here, which has a loading state (no arguments), an error state (with an error), and a data state with a (Fido|Oath)State. One solution could be to make it a type of error (eg. NotInitializedError) that can have special handling.

Reply:
Good idea, I will update this to be some kind of error state 👍🏻

Reply:
Fixed!